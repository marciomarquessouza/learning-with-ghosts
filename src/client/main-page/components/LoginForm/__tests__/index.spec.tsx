import React from 'react'

import { render, fireEvent } from '@testing-library/react'

import LoginFrom from '..'

describe('components >> LoginForm', () => {
    const defaultMockedProps = {
        email: '__MOCK__EMAIL__',
        password: '__MOCK__PASSWORD__',
        loading: false,
        onChangeEmail: jest.fn(),
        onChangePassword: jest.fn(),
        onClickRegister: jest.fn(),
        onClickForgotPassword: jest.fn(),
        onClickTwitterLogin: jest.fn(),
        onClickGoogleLogin: jest.fn(),
        onSubmit: jest.fn(),
    }

    it('should render the LoginFrom properly', () => {
        const { container } = render(<LoginFrom {...defaultMockedProps} />)

        expect(container).toMatchSnapshot('LoginFrom-default')
    })

    it('should render the LoginFrom properly in loading state', () => {
        const mockProps = { ...defaultMockedProps, loading: true }
        const { container } = render(<LoginFrom {...mockProps} />)

        expect(container).toMatchSnapshot('LoginFrom-loading')
    })

    it('should fire the onSubmit event when the button submit is pressed', () => {
        const spyOnSubmit = jest.fn()
        const mockedProps = {
            ...defaultMockedProps,
            onSubmit: spyOnSubmit,
        }
        const { getByText } = render(<LoginFrom {...mockedProps} />)

        fireEvent.click(getByText('Login'))

        expect(spyOnSubmit).toHaveBeenCalledTimes(1)
    })

    it('should fire the onClickGoogleLogin event when the google icon is pressed', () => {
        const spyOnClickGoogleLogin = jest.fn()
        const mockedProps = {
            ...defaultMockedProps,
            onClickGoogleLogin: spyOnClickGoogleLogin,
        }
        const { getByRole } = render(<LoginFrom {...mockedProps} />)

        fireEvent.click(getByRole('button', { name: /google/i }))

        expect(spyOnClickGoogleLogin).toHaveBeenCalledTimes(1)
    })

    it('should fire the onClickTwitterLogin event when the twitter icon is pressed', () => {
        const spyOnClickTwitterLogin = jest.fn()
        const mockedProps = {
            ...defaultMockedProps,
            onClickTwitterLogin: spyOnClickTwitterLogin,
        }
        const { getByRole } = render(<LoginFrom {...mockedProps} />)
        fireEvent.click(getByRole('button', { name: /twitter/i }))

        expect(spyOnClickTwitterLogin).toHaveBeenCalledTimes(1)
    })

    it('should fire the onClickForgotPassword event when the forgot password link is pressed', () => {
        const spyOnClickForgotPassword = jest.fn()
        const mockedProps = {
            ...defaultMockedProps,
            onClickForgotPassword: spyOnClickForgotPassword,
        }
        const { getByText } = render(<LoginFrom {...mockedProps} />)
        fireEvent.click(getByText('Forgot password?'))

        expect(spyOnClickForgotPassword).toHaveBeenCalledTimes(1)
    })

    it('should fire the onClickRegister event when the register link is pressed', () => {
        const spyOnClickRegister = jest.fn()
        const mockedProps = {
            ...defaultMockedProps,
            onClickRegister: spyOnClickRegister,
        }
        const { getByText } = render(<LoginFrom {...mockedProps} />)
        fireEvent.click(getByText('Register'))

        expect(spyOnClickRegister).toHaveBeenCalledTimes(1)
    })

    it('should fire the onChangeEmail event when the email is entered by the user', () => {
        const spyOnChangeEmail = jest.fn()
        const mockEmail = '__MOCK__EMAIL__'
        const mockedProps = {
            ...defaultMockedProps,
            email: '',
            onChangeEmail: spyOnChangeEmail,
        }
        const { getByPlaceholderText } = render(<LoginFrom {...mockedProps} />)
        fireEvent.change(getByPlaceholderText('Email address'), { target: { value: mockEmail } })

        expect(spyOnChangeEmail).toHaveBeenCalledTimes(1)
        expect(spyOnChangeEmail).toHaveBeenCalledWith(mockEmail)
    })

    it('should fire the onChangePassword event when the password is entered by the user', () => {
        const spyOnChangePassword = jest.fn()
        const mockPassword = '__MOCK__PASSWORD__'
        const mockedProps = {
            ...defaultMockedProps,
            password: '',
            onChangePassword: spyOnChangePassword,
        }
        const { getByPlaceholderText } = render(<LoginFrom {...mockedProps} />)
        fireEvent.change(getByPlaceholderText('Password'), { target: { value: mockPassword } })

        expect(spyOnChangePassword).toHaveBeenCalledTimes(1)
        expect(spyOnChangePassword).toHaveBeenCalledWith(mockPassword)
    })
})
