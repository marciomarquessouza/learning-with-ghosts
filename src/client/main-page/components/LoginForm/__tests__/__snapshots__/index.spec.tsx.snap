// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components >> LoginForm should render the LoginFrom properly in loading state: LoginFrom-loading 1`] = `
<div>
  <form>
    <div
      class="flex flex-row items-center justify-center lg:justify-start"
    >
      <p
        class="text-3xl text-white mb-0 mr-4"
      >
        Looogin
      </p>
      <button
        aria-label="twitter"
        class="inline-block p-3 bg-primary-light text-white font-medium text-xs leading-tight uppercase rounded-full shadow-md hover:bg-primary-dark hover:shadow-lg focus:bg-primary focus:shadow-lg focus:outline-none focus:ring-0 active:bg-primary active:shadow-lg transition duration-150 ease-in-out mx-1"
        data-mdb-ripple="true"
        data-mdb-ripple-color="light"
        name="twitter"
        type="button"
      >
        <svg
          class="w-4 h-4"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="google"
        class="inline-block p-3 bg-primary-light text-white font-medium text-xs leading-tight uppercase rounded-full shadow-md hover:bg-primary-dark hover:shadow-lg focus:bg-primary focus:shadow-lg focus:outline-none focus:ring-0 active:bg-primary active:shadow-lg transition duration-150 ease-in-out mx-1"
        data-mdb-ripple="true"
        data-mdb-ripple-color="light"
        name="google"
        type="button"
      >
        <svg
          class="w-4 h-4"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M881 442.4H519.7v148.5h206.4c-8.9 48-35.9 88.6-76.6 115.8-34.4 23-78.3 36.6-129.9 36.6-99.9 0-184.4-67.5-214.6-158.2-7.6-23-12-47.6-12-72.9s4.4-49.9 12-72.9c30.3-90.6 114.8-158.1 214.7-158.1 56.3 0 106.8 19.4 146.6 57.4l110-110.1c-66.5-62-153.2-100-256.6-100-149.9 0-279.6 86-342.7 211.4-26 51.8-40.8 110.4-40.8 172.4S151 632.8 177 684.6C240.1 810 369.8 896 519.7 896c103.6 0 190.4-34.4 253.8-93 72.5-66.8 114.4-165.2 114.4-282.1 0-27.2-2.4-53.3-6.9-78.5z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="flex items-center justify-center my-4 before:flex-1 before:border-t before:border-gray-300 before:mt-0.5 after:flex-1 after:border-t after:border-gray-300 after:mt-0.5"
    >
      <p
        class="text-white text-center font-semibold mx-4 mb-0"
      >
        Or
      </p>
    </div>
    <div
      class="mb-6"
    >
      <input
        class="form-control block w-full px-4 py-2 text-xl font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
        name="email"
        placeholder="Email address"
        type="text"
        value="__MOCK__EMAIL__"
      />
    </div>
    <div
      class="mb-6"
    >
      <input
        class="form-control block w-full px-4 py-2 text-xl font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
        name="password"
        placeholder="Password"
        type="password"
        value="__MOCK__PASSWORD__"
      />
    </div>
    <div
      class="flex items-center justify-center md:justify-start mb-6"
    >
      <a
        class="text-white"
        href="#"
      >
        Forgot password?
      </a>
    </div>
    <div
      class="text-center lg:text-left"
    >
      <button
        class="inline-block px-7 py-3 bg-primary-light text-white font-medium text-sm leading-snug uppercase rounded shadow-md hover:bg-primary-dark hover:shadow-lg focus:bg-primary focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-800 active:shadow-lg transition duration-150 ease-in-out"
        type="submit"
      >
        <div
          class="flex items-center justify-center"
        >
          <div
            class="spinner-border animate-spin inline-block w-8 h-8 border-4 rounded-full"
            role="status"
          >
            <span
              class="visually-hidden"
            >
              Loading...
            </span>
          </div>
        </div>
      </button>
      <p
        class="text-lg text-white font-semibold mt-4 pt-1 mb-0"
      >
        Don't have an account?
         
        <a
          class="text-primary-light hover:text-primary-dark focus:text-primary transition duration-200 ease-in-out"
          href="#"
        >
          Register
        </a>
      </p>
    </div>
  </form>
</div>
`;

exports[`components >> LoginForm should render the LoginFrom properly: LoginFrom-default 1`] = `
<div>
  <form>
    <div
      class="flex flex-row items-center justify-center lg:justify-start"
    >
      <p
        class="text-3xl text-white mb-0 mr-4"
      >
        Looogin
      </p>
      <button
        aria-label="twitter"
        class="inline-block p-3 bg-primary-light text-white font-medium text-xs leading-tight uppercase rounded-full shadow-md hover:bg-primary-dark hover:shadow-lg focus:bg-primary focus:shadow-lg focus:outline-none focus:ring-0 active:bg-primary active:shadow-lg transition duration-150 ease-in-out mx-1"
        data-mdb-ripple="true"
        data-mdb-ripple-color="light"
        name="twitter"
        type="button"
      >
        <svg
          class="w-4 h-4"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="google"
        class="inline-block p-3 bg-primary-light text-white font-medium text-xs leading-tight uppercase rounded-full shadow-md hover:bg-primary-dark hover:shadow-lg focus:bg-primary focus:shadow-lg focus:outline-none focus:ring-0 active:bg-primary active:shadow-lg transition duration-150 ease-in-out mx-1"
        data-mdb-ripple="true"
        data-mdb-ripple-color="light"
        name="google"
        type="button"
      >
        <svg
          class="w-4 h-4"
          viewBox="0 0 1024 1024"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M881 442.4H519.7v148.5h206.4c-8.9 48-35.9 88.6-76.6 115.8-34.4 23-78.3 36.6-129.9 36.6-99.9 0-184.4-67.5-214.6-158.2-7.6-23-12-47.6-12-72.9s4.4-49.9 12-72.9c30.3-90.6 114.8-158.1 214.7-158.1 56.3 0 106.8 19.4 146.6 57.4l110-110.1c-66.5-62-153.2-100-256.6-100-149.9 0-279.6 86-342.7 211.4-26 51.8-40.8 110.4-40.8 172.4S151 632.8 177 684.6C240.1 810 369.8 896 519.7 896c103.6 0 190.4-34.4 253.8-93 72.5-66.8 114.4-165.2 114.4-282.1 0-27.2-2.4-53.3-6.9-78.5z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="flex items-center justify-center my-4 before:flex-1 before:border-t before:border-gray-300 before:mt-0.5 after:flex-1 after:border-t after:border-gray-300 after:mt-0.5"
    >
      <p
        class="text-white text-center font-semibold mx-4 mb-0"
      >
        Or
      </p>
    </div>
    <div
      class="mb-6"
    >
      <input
        class="form-control block w-full px-4 py-2 text-xl font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
        name="email"
        placeholder="Email address"
        type="text"
        value="__MOCK__EMAIL__"
      />
    </div>
    <div
      class="mb-6"
    >
      <input
        class="form-control block w-full px-4 py-2 text-xl font-normal text-gray-700 bg-white bg-clip-padding border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
        name="password"
        placeholder="Password"
        type="password"
        value="__MOCK__PASSWORD__"
      />
    </div>
    <div
      class="flex items-center justify-center md:justify-start mb-6"
    >
      <a
        class="text-white"
        href="#"
      >
        Forgot password?
      </a>
    </div>
    <div
      class="text-center lg:text-left"
    >
      <button
        class="inline-block px-7 py-3 bg-primary-light text-white font-medium text-sm leading-snug uppercase rounded shadow-md hover:bg-primary-dark hover:shadow-lg focus:bg-primary focus:shadow-lg focus:outline-none focus:ring-0 active:bg-blue-800 active:shadow-lg transition duration-150 ease-in-out"
        type="submit"
      >
        Login
      </button>
      <p
        class="text-lg text-white font-semibold mt-4 pt-1 mb-0"
      >
        Don't have an account?
         
        <a
          class="text-primary-light hover:text-primary-dark focus:text-primary transition duration-200 ease-in-out"
          href="#"
        >
          Register
        </a>
      </p>
    </div>
  </form>
</div>
`;
